### DISCUSSION TITLE
using AdaptiveCpp on AMDGPU via SPIR-V/OpenCL backend?

### DISCUSSION THREAD

**Discussion by BenWibking** [AdaptiveCpp/AdaptiveCpp#1564] [Q&A] [answered]
Is it possible to use the generic single-pass mode to run on AMD GPUs using the OpenCL backend?

The motivation for doing this is to avoid the LLVM AMDGPU code generation backend, which generates incorrect register spill code for kernels with high register pressure and divergent execution (see, e.g.: https://discourse.llvm.org/t/the-current-state-of-spilling-function-calls-and-related-problems/2863 -- this is still the case today, since LLVM register allocation does not understand predicated execution).

SPIR-V kernels can be compiled to machine code with the Mesa ACO compiler, which does not have this problem: https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/amd/compiler/README.md

**Comment by illuhad** [comment, marked as answer]
This is currently not possible because there is no OpenCL implementation for AMD that supports both SPIR-V and USM, which is needed for SYCL (or any modern single-source programming model). As far as I know rusticl is working towards this, but is still a long shot away from being usable here.
